File: r/import_data/import_bbref.R
library(tidyr)
library(stringr)
library(dplyr)
library(stringdist)


bbref_batters_2020 <- read.csv("~/Documents/bb/2021/data/bbref_batters_2020.csv", stringsAsFactors=FALSE)

bbref_batters <- bbref_batters_2020

# drop Rank
bbref_batters <- data.frame(bbref_batters %>% select(-Rk))

# Clean up the names
bbref_batters <- data.frame(bbref_batters %>% separate(Name, c("name", "bbref_id"), sep = "\\\\"))
bbref_batters$name <- str_replace(bbref_batters$name, '\\*', "")
bbref_batters$name <- str_replace(bbref_batters$name, '\\#', "")

# Filter to only have annual totals
num_obs <- data.frame(bbref_batters %>% count(name, sort=TRUE))
bbref_batters = merge(bbref_batters, num_obs, by="name") %>% subset(n==1 | Tm=='TOT') %>% select(-'n')
num_obs <- data.frame(bbref_batters %>% count(name, sort=TRUE))
bbref_batters = merge(bbref_batters, num_obs, by="name") %>% subset(n==1 | Lg=='MLB') %>% select(-'n')



rm(bbref_batters_2020, num_obs)

File: python/munging/import_bbref.py
def import_bbref(filename):
    # filename = 'data/bbref_qs_historical.csv'
    import pandas as pd

    df = pd.read_csv(filename)

    del(df['Rk'])

    if ('Player' in df.columns):
        df['Name'] = df['Player']
        del(df['Player'])

    df[['name','bbref_id']] = df['Name'].str.split('\\', expand=True)
    df[['name']] = df['name'].str.replace('\*','', regex=True)
    df[['name']] = df['name'].str.replace('\#','', regex=True)
    del(df['Name'])

    # Filter to only have annual totals
    df[['has_tot']] = df['Tm']=='TOT'
    df[['has_tot']] = df.groupby('bbref_id')['has_tot'].transform('max')
    df = df.drop(df[(df['has_tot']==1) & (df['Tm']!='TOT')].index)
    del(df['has_tot'])

    if ('Lg' in df.columns):
        df[['tot_mlb']] = (df['Tm']=='TOT') & (df['Lg']=='MLB')
        df[['tot_mlb']] = df.groupby('bbref_id')['tot_mlb'].transform('max')
        df = df.drop(df[(df['tot_mlb']==1) & (df['Lg']!='MLB')].index)
        del(df['tot_mlb'])


    colnames = df.columns.tolist()
    colnames.remove('name')
    colnames.insert(0, 'name')
    colnames.remove('bbref_id')
    colnames.insert(1, 'bbref_id')

    df = df.reindex(columns=colnames)
    return df


#bbref_batters = import_bbref('data/bbref_batters_2020.csv')
#bbref_pitchers = import_bbref('data/bbref_pitchers_2020.csv')
#bbref_fielding_2020 = import_bbref('data/bbref_fielding_appearances_2020.csv')
#bbref_qs = import_bbref('data/bbref_qs_historical.csv')

